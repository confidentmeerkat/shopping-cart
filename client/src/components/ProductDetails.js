import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import axios from 'axios';
import numeral from 'numeral';
import RaisedButton from 'material-ui/RaisedButton';
import Snackbar from 'material-ui/Snackbar';
import AddShoppingCart from 'material-ui/svg-icons/action/add-shopping-cart';
import KeyboardArrowLeft from 'material-ui/svg-icons/hardware/keyboard-arrow-left';
import '../styles/ProductDetails.css';

class ProductDetails extends React.Component {

  state = {
    postData: {
      user: this.props.loggedUser._id,
      product: this.props.product._id,
      quantity: 1
    },
    snackbarOpen: false
  }

  onQuantityChange = (e) => {
    let value = e.target.value;
    this.setState((prevState) => ({
      postData: { ...prevState.postData, quantity: value*1 }
    }));
  }

  addToCart = () => {
    axios.post('/api/cart', this.state.postData);
    this.setState({ snackbarOpen: true });
  }

  render() {
    return (
      <div className="product-details-container">
        <h1>{this.props.product.info.name}</h1>
        <div className="product-details">
          <div className="product-image">
            <img src={this.props.product.info.photo} />
          </div>
          <div className="product-info">
            <table>
              <tr>
                <th>Model</th>
                <td>{this.props.product.info.name}</td>
              </tr>
              <tr>
                <th>Dimensions</th>
                <td>{this.props.product.info.dimensions}</td>
              </tr>
              <tr>
                <th>Weight</th>
                <td>{this.props.product.info.weight}</td>
              </tr>
              <tr>
                <th>Display Type</th>
                <td>{this.props.product.info.displayType}</td>
              </tr>
              <tr>
                <th>Display Size</th>
                <td>{this.props.product.info.displaySize}</td>
              </tr>
              <tr>
                <th>Display Resolution</th>
                <td>{this.props.product.info.displayResolution}</td>
              </tr>
              <tr>
                <th>OS</th>
                <td>{this.props.product.info.os}</td>
              </tr>
              <tr>
                <th>CPU</th>
                <td>{this.props.product.info.cpu}</td>
              </tr>
              <tr>
                <th>Internal Memory</th>
                <td>{this.props.product.info.internalMemory}</td>
              </tr>
              <tr>
                <th>RAM</th>
                <td>{this.props.product.info.ram}</td>
              </tr>
              <tr>
                <th>Camera</th>
                <td>{this.props.product.info.camera}</td>
              </tr>
              <tr>
                <th>Batery</th>
                <td>{this.props.product.info.batery}</td>
              </tr>
              <tr>
                <th>Color</th>
                <td>{this.props.product.info.color}</td>
              </tr>
            </table>
            <Snackbar
              open={this.state.snackbarOpen}
              message={this.props.loggedUser ? 'Item added to your cart.' : 'You must be logged in!'}
              autoHideDuration={4000}
              bodyStyle={this.props.loggedUser ? { 'background': '#64DD17' } : { 'background': '#F44336' }}
            />
          </div>
        </div>
        <div className="product-handle">
          <div className="left">
            <RaisedButton
              containerElement={<Link to="/" />}
              className="btn"
              label="Back to catalog"
              labelPosition="after"
              secondary={true}
              icon={<KeyboardArrowLeft />}
            />
          </div>
          <div className="right">
            <span className="price-text">Price: </span>
            <span className="price-num">{numeral(this.props.product.info.price).format('$0,0.00')}</span>
            <span className="price-text">Quantity: </span>
            <span><input type="number" value={this.state.postData.quantity} min="1" max="5" onChange={this.onQuantityChange} /></span>
            <RaisedButton
              onClick={this.addToCart}
              label="Add to cart"
              labelPosition="before"
              primary={true}
              icon={<AddShoppingCart />}
            />
           </div> 
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  loggedUser: state.loggedUser,
  product: state.catalog.find((item) => item._id == props.match.params.id)
});

export default connect(mapStateToProps)(ProductDetails);