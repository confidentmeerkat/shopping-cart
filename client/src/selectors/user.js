export const selectUser = (state) => state.loggedUser.user;