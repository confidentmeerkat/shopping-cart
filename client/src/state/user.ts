import { Order } from './order';

export interface User {
  orders: Order[];
  username: string;
  email: string;
  address: string;
  phone: string;
  _id: string;
}
