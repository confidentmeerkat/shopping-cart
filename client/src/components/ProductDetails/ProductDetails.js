import React from 'react';
import { Link } from 'react-router-dom';
import axios from 'axios';
import numeral from 'numeral';
import RaisedButton from 'material-ui/RaisedButton';
import Snackbar from 'material-ui/Snackbar';
import AddShoppingCart from 'material-ui/svg-icons/action/add-shopping-cart';
import KeyboardArrowLeft from 'material-ui/svg-icons/hardware/keyboard-arrow-left';
import '@styles/ProductDetails.css';

export class ProductDetails extends React.Component {
  state = {
    postData: {
      user: this.props.loggedUser._id,
      product: this.props.product._id,
      quantity: 1
    },
    snackbarOpen: false
  }

  onQuantityChange = (e) => {
    let value = e.target.value;
    this.setState((prevState) => ({
      postData: { ...prevState.postData, quantity: value*1 }
    }));
  }

  addToCart = () => {
    axios.post('/api/cart', this.state.postData);
    this.setState({ snackbarOpen: true });
  }

  render() {
    const {
      loggedUser,
      product: {
        info: {
          name,
          photo,
          dimensions,
          weight,
          displayType,
          displaySize,
          displayResolution,
          os,
          cpu,
          internalMemory,
          ram,
          camera,
          batery,
          color,
          price
        }
      }
    } = this.props;

    return (
      <div className="product-details-container">
        <h1>{name}</h1>
        <div className="product-details">
          <div className="product-image">
            <img src={photo} />
          </div>
          <div className="product-info">
            <table>
              <tr>
                <th>Model</th>
                <td>{name}</td>
              </tr>
              <tr>
                <th>Dimensions</th>
                <td>{dimensions}</td>
              </tr>
              <tr>
                <th>Weight</th>
                <td>{weight}</td>
              </tr>
              <tr>
                <th>Display Type</th>
                <td>{displayType}</td>
              </tr>
              <tr>
                <th>Display Size</th>
                <td>{displaySize}</td>
              </tr>
              <tr>
                <th>Display Resolution</th>
                <td>{displayResolution}</td>
              </tr>
              <tr>
                <th>OS</th>
                <td>{os}</td>
              </tr>
              <tr>
                <th>CPU</th>
                <td>{cpu}</td>
              </tr>
              <tr>
                <th>Internal Memory</th>
                <td>{internalMemory}</td>
              </tr>
              <tr>
                <th>RAM</th>
                <td>{ram}</td>
              </tr>
              <tr>
                <th>Camera</th>
                <td>{camera}</td>
              </tr>
              <tr>
                <th>Batery</th>
                <td>{batery}</td>
              </tr>
              <tr>
                <th>Color</th>
                <td>{color}</td>
              </tr>
            </table>
            <Snackbar
              open={this.state.snackbarOpen}
              message={loggedUser ? 'Item added to your cart.' : 'You must be logged in!'}
              autoHideDuration={4000}
              bodyStyle={loggedUser ? { 'background': '#64DD17' } : { 'background': '#F44336' }}
            />
          </div>
        </div>
        <div className="product-handle">
          <div className="left">
            <RaisedButton
              containerElement={<Link to="/" />}
              className="btn"
              label="Back to catalog"
              labelPosition="after"
              secondary={true}
              icon={<KeyboardArrowLeft />}
            />
          </div>
          <div className="right">
            <div className="price">
              <span className="price-text">Price: </span>
              <span className="price-num">{numeral(price).format('$0,0.00')}</span>
            </div>
            <div className="quantity">
              <span className="price-text">Quantity: </span>
              <span><input type="number" value={this.state.postData.quantity} min="1" max="5" onChange={this.onQuantityChange} /></span>
            </div>
            <div className="btn">
              <RaisedButton
                onClick={this.addToCart}
                label="Add to cart"
                labelPosition="before"
                primary={true}
                icon={<AddShoppingCart />}
              />
            </div>
           </div> 
        </div>
      </div>
    );
  }
}

export default ProductDetails;