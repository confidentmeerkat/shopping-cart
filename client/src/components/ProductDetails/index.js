import { compose } from 'recompose';
import { connect } from 'react-redux';
import ProductDetails from './ProductDetails';

const mapStateToProps = (state, props) => ({
  loggedUser: state.loggedUser.user,
  product: state.catalog.items.find((item) => item._id == props.match.params.id)
});

export default compose(
  connect(mapStateToProps)
)(ProductDetails);
