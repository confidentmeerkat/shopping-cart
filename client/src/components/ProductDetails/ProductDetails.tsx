import * as React from 'react';
import { Link } from 'react-router-dom';
import axios from 'axios';
import * as numeral from 'numeral';
import RaisedButton from 'material-ui/RaisedButton';
import Snackbar from 'material-ui/Snackbar';
import AddShoppingCart from 'material-ui/svg-icons/action/add-shopping-cart';
import KeyboardArrowLeft from 'material-ui/svg-icons/hardware/keyboard-arrow-left';
import { IUser, ICatalogProduct } from '@typings/state/index';
import '@styles/ProductDetails.css';

interface Props {
  loggedUser: IUser;
  product: ICatalogProduct;
}

interface State {
  postData: {
    user: string;
    product: string;
    quantity: number;
  };
  snackbarOpen: boolean;
}

class ProductDetails extends React.Component<Props, State> {
  state = {
    postData: {
      user: this.props.loggedUser._id,
      product: this.props.product._id,
      quantity: 1
    },
    snackbarOpen: false
  }

  onQuantityChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    let value = e.target.value;

    this.setState((prevState: State) => ({
      postData: { ...prevState.postData, quantity: parseInt(value) }
    }));
  }

  addToCart = () => {
    axios.post('/api/cart', this.state.postData);

    this.setState({ snackbarOpen: true });
  }

  render() {
    const {
      loggedUser,
      product: { info }
    } = this.props;

    return (
      <div className="product-details-container">
        <h1>{info.name}</h1>
        <div className="product-details">
          <div className="product-image">
            <img src={info.photo} />
          </div>
          <div className="product-info">
            <table>
              <tr>
                <th>Model</th>
                <td>{info.name}</td>
              </tr>
              <tr>
                <th>Dimensions</th>
                <td>{info.dimensions}</td>
              </tr>
              <tr>
                <th>Weight</th>
                <td>{info.weight}</td>
              </tr>
              <tr>
                <th>Display Type</th>
                <td>{info.displayType}</td>
              </tr>
              <tr>
                <th>Display Size</th>
                <td>{info.displaySize}</td>
              </tr>
              <tr>
                <th>Display Resolution</th>
                <td>{info.displayResolution}</td>
              </tr>
              <tr>
                <th>OS</th>
                <td>{info.os}</td>
              </tr>
              <tr>
                <th>CPU</th>
                <td>{info.cpu}</td>
              </tr>
              <tr>
                <th>Internal Memory</th>
                <td>{info.internalMemory}</td>
              </tr>
              <tr>
                <th>RAM</th>
                <td>{info.ram}</td>
              </tr>
              <tr>
                <th>Camera</th>
                <td>{info.camera}</td>
              </tr>
              <tr>
                <th>Batery</th>
                <td>{info.batery}</td>
              </tr>
              <tr>
                <th>Color</th>
                <td>{info.color}</td>
              </tr>
            </table>
            <Snackbar
              open={this.state.snackbarOpen}
              message={loggedUser ? 'Item added to your cart.' : 'You must be logged in!'}
              autoHideDuration={4000}
              bodyStyle={loggedUser ? { 'background': '#64DD17' } : { 'background': '#F44336' }}
            />
          </div>
        </div>
        <div className="product-handle">
          <div className="left">
            <RaisedButton
              containerElement={<Link to="/" />}
              className="btn"
              label="Back to catalog"
              labelPosition="after"
              secondary={true}
              icon={<KeyboardArrowLeft />}
            />
          </div>
          <div className="right">
            <div className="price">
              <span className="price-text">Price: </span>
              <span className="price-num">{numeral(info.price).format('$0,0.00')}</span>
            </div>
            <div className="quantity">
              <span className="price-text">Quantity: </span>
              <span><input type="number" value={this.state.postData.quantity} min="1" max="5" onChange={this.onQuantityChange} /></span>
            </div>
            <div className="btn">
              <RaisedButton
                onClick={this.addToCart}
                label="Add to cart"
                labelPosition="before"
                primary={true}
                icon={<AddShoppingCart />}
              />
            </div>
           </div>
        </div>
      </div>
    );
  }
}

export default ProductDetails;
