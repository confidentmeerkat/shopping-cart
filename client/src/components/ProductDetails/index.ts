import { compose } from 'recompose';
import { connect } from 'react-redux';
import { selectUser } from '@selectors/user';
import { selectProduct } from '@selectors/catalog';
import { IState } from '@typings/state/index'
import ProductDetails from './ProductDetails';

const mapStateToProps = (state: IState, ownProps: any) => ({
  loggedUser: selectUser(state),
  product: selectProduct(state, ownProps)
});

export default compose(
  connect(mapStateToProps)
)(ProductDetails);
