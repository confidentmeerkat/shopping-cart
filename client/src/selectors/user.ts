import { IState, IUser } from '@typings/state/index';

export const selectUser = (state: IState): IUser | null => state.loggedUser.user;
