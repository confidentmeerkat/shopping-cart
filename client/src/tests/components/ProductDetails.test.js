import React from 'react';
import { shallow } from 'enzyme';
import { ProductDetails } from '../../components/ProductDetails';
import user from '../__mocks__/user';
import item from '../__mocks__/item';

describe('<ProductDetails />', () => {
  it('renders ProductDetails correctly', () => {
    const wrapper = shallow(<ProductDetails loggedUser={user} product={item} />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('h1').text()).toBe('Samsung Galaxy S8');
    expect(wrapper.find('img').prop('src')).toBe('/img/samsung_galaxy_s8.jpg');
    expect(wrapper.find('table').children()).toHaveLength(13);
    expect(wrapper.find('.price').text()).toBe('Price: $720.00');
  });

  it('handles add to cart button when no user is logged in', () => {
    const wrapper = shallow(<ProductDetails loggedUser="" product={item} />);

    wrapper.find('RaisedButton').at(1).simulate('click');
    expect(wrapper.state('snackbarOpen')).toBe(true);
    expect(wrapper.find('Snackbar').prop('message')).toBe('You must be logged in!');
  });

  it('handles add to cart button when user is logged in', () => {
    const wrapper = shallow(<ProductDetails loggedUser={user} product={item} />);

    wrapper.find('RaisedButton').at(1).simulate('click');
    expect(wrapper.state('snackbarOpen')).toBe(true);
    expect(wrapper.find('Snackbar').prop('message')).toBe('Item added to your cart.');
  });
});