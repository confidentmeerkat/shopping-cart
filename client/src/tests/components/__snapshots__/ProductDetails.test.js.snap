// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductDetails /> renders ProductDetails correctly 1`] = `
<div
  className="product-details-container"
>
  <h1>
    Samsung Galaxy S8
  </h1>
  <div
    className="product-details"
  >
    <div
      className="product-image"
    >
      <img
        src="/img/samsung_galaxy_s8.jpg"
      />
    </div>
    <div
      className="product-info"
    >
      <table>
        <tr>
          <th>
            Model
          </th>
          <td>
            Samsung Galaxy S8
          </td>
        </tr>
        <tr>
          <th>
            Dimensions
          </th>
          <td>
            148.9 x 68.1 x 8 mm
          </td>
        </tr>
        <tr>
          <th>
            Weight
          </th>
          <td>
            155 g
          </td>
        </tr>
        <tr>
          <th>
            Display Type
          </th>
          <td>
            Super AMOLED capacitive touchscreen, 16M colors
          </td>
        </tr>
        <tr>
          <th>
            Display Size
          </th>
          <td>
            5.8"
          </td>
        </tr>
        <tr>
          <th>
            Display Resolution
          </th>
          <td>
            1440 x 2960 pixels
          </td>
        </tr>
        <tr>
          <th>
            OS
          </th>
          <td>
            Android 7.0 (Nougat)
          </td>
        </tr>
        <tr>
          <th>
            CPU
          </th>
          <td>
            Octa-core (4x2.3 GHz & 4x1.7 GHz) - EMEA
          </td>
        </tr>
        <tr>
          <th>
            Internal Memory
          </th>
          <td>
            64 GB
          </td>
        </tr>
        <tr>
          <th>
            RAM
          </th>
          <td>
            4 GB
          </td>
        </tr>
        <tr>
          <th>
            Camera
          </th>
          <td>
            12 MP (f/1.7, 26mm, 1/2.5", 1.4 Âµm, Dual Pixel PDAF), phase detection autofocus, OIS
          </td>
        </tr>
        <tr>
          <th>
            Batery
          </th>
          <td>
            Non-removable Li-Ion 3000 mAh battery
          </td>
        </tr>
        <tr>
          <th>
            Color
          </th>
          <td>
            Midnight Black
          </td>
        </tr>
      </table>
      <Snackbar
        autoHideDuration={4000}
        bodyStyle={
          Object {
            "background": "#64DD17",
          }
        }
        message="Item added to your cart."
        open={false}
      />
    </div>
  </div>
  <div
    className="product-handle"
  >
    <div
      className="left"
    >
      <RaisedButton
        className="btn"
        containerElement={
          <Link
            replace={false}
            to="/"
          />
        }
        disabled={false}
        fullWidth={false}
        icon={<HardwareKeyboardArrowLeft />}
        label="Back to catalog"
        labelPosition="after"
        primary={false}
        secondary={true}
      />
    </div>
    <div
      className="right"
    >
      <div
        className="price"
      >
        <span
          className="price-text"
        >
          Price: 
        </span>
        <span
          className="price-num"
        >
          $720.00
        </span>
      </div>
      <div
        className="quantity"
      >
        <span
          className="price-text"
        >
          Quantity: 
        </span>
        <span>
          <input
            max="5"
            min="1"
            onChange={[Function]}
            type="number"
            value={1}
          />
        </span>
      </div>
      <div
        className="btn"
      >
        <RaisedButton
          disabled={false}
          fullWidth={false}
          icon={<ActionAddShoppingCart />}
          label="Add to cart"
          labelPosition="before"
          onClick={[Function]}
          primary={true}
          secondary={false}
        />
      </div>
    </div>
  </div>
</div>
`;
